modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java

```diff
@@ -136,14 +162,29 @@ public void onMessage(NetworkMessage message) {
             this.remoteLaunchId = msg.launchId();
             this.remoteConsistentId = msg.consistentId();
 
-            this.recoveryDescriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
-                    remoteConsistentId,
-                    remoteLaunchId,
-                    connectionId,
-                    false
-            );
+            channelCreationListener.notifyInboundChannelCreation(this.remoteConsistentId, this.connectionId)
+                    .whenCompleteAsync((unused, throwable) -> {
+                        RecoveryDescriptor descriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
+                                remoteConsistentId,
+                                remoteLaunchId,
+                                connectionId,
+                                false
+                        );
+
+                        if (!descriptor.acquire(ctx)) {
+                            String err = "Failed to acquire recovery descriptor during handshake, it is held by: " + descriptor.holder();
+
+                            LOG.warn(err);
+
+                            handshakeCompleteFuture.completeExceptionally(new HandshakeException(err));
```

Comment:
Maybe you should also close the connection or send some response, like "already opened"

Reply:
Done